R/convenience.R

```diff
@@ -129,45 +129,106 @@ LoadNanostring <- function(data.dir, fov, assay = 'Nanostring') {
+#' @param update.object If to update final object, default to TRUE.
+#' @param ... Arguments passed to \code{ReadVizgen}
+#'
 #' @importFrom SeuratObject Cells CreateCentroids CreateFOV
 #' CreateSegmentation CreateSeuratObject
+#' @import dplyr
 #'
 #' @export
 #'
 #' @rdname ReadVizgen
 #'
-LoadVizgen <- function(data.dir, fov, assay = 'Vizgen', z = 3L) {
-  data <- ReadVizgen(
-    data.dir = data.dir,
-    filter = "^Blank-",
-    type = c("centroids", "segmentations"),
-    z = z
-  )
-  segs <- CreateSegmentation(data$segmentations)
-  cents <- CreateCentroids(data$centroids)
-  segmentations.data <- list(
-    "centroids" = cents,
-    "segmentation" = segs
-  )
-  coords <- CreateFOV(
-    coords = segmentations.data,
-    type = c("segmentation", "centroids"),
-    molecules = data$microns,
-    assay = assay
-  )
-  obj <- CreateSeuratObject(counts = data$transcripts, assay = assay)
-  # only consider the cells we have counts and a segmentation for
-  # Cells which don't have a segmentation are probably found in other z slices.
-  coords <- subset(
-    x = coords,
-    cells = intersect(
-      x = Cells(x = coords[["segmentation"]]),
-      y = Cells(x = obj)
-    )
-  )
-  # add coords to seurat object
+LoadVizgen <- function(data.dir, fov = "vz", assay = 'Vizgen',
+                       add.zIndex = TRUE, update.object = TRUE,
+                       ...) {
+  data <- ReadVizgen(data.dir = data.dir, ...)
+  
+  #gc() %>% invisible()
+  message("Creating Seurat object..")  
+  obj <- CreateSeuratObject(counts = data[["transcripts"]], assay = assay)
+  
+  # in case no segmentation is present, use boxes
+  if (!"segmentations" %in% names(data)) {
+    if ("boxes" %in% names(data)) {
+      bound.boxes <- CreateSegmentation(data[["boxes"]])
+      cents <- CreateCentroids(data[["centroids"]])
+      bound.boxes.data <- list(centroids = cents, 
+                               boxes = bound.boxes)
+      message("Creating FOVs..", "\n",
+              ">>> using box coordinates instead of segmentations")
+      coords <- CreateFOV(coords = bound.boxes.data, 
+                          type = c("boxes", "centroids"), 
+                          molecules = data[[mol.type]], 
+                          assay = assay)
+    } else { # in case no segmentation & no boxes are present, use centroids only
+      cents <- CreateCentroids(data[["centroids"]])
+      message("Creating FOVs..", "\n",
+              ">>> using only centroids")
+      coords <- CreateFOV(coords = list(centroids = cents), 
+                          type = c("centroids"),
+                          molecules = data[[mol.type]], 
+                          assay = assay)
+    }
+    # only consider the cells we have counts and a segmentation for
+    # Cells which don't have a segmentation are probably found in other z slices.
+    coords <- subset(x = coords, 
+                     cells = intersect(x = Cells(x = coords[["boxes"]]),
```

Comment:
coords[['boxes']] needs to be changes to coords[['centroids']] to account for situations where type = 'centroids' (which wouldn't have coords[['boxes']])

Reply:
yes, actually working on a 2nd large fix for a number of things, including keeping counts `> 0`

Reply:
great! I'm at Vizgen and testing this PR on a number of public and internal datasets

Reply:
Thanks for all the great work!